### PR TITLE
Reverse order of quota application

### PR DESCRIPTION
**Pull request by scrosby**
## Changes proposed in this PR

- Reverse order of quota application

## Why are we making these changes?

Applying quotas in the wrong order can also cause us to not launch
jobs that are within quota. A user who's already at their quota will
appear to falsely use global quota that they can't really use.

### CODE REVIEW

File: scheduler/src/cook/tools.clj [hunk 1]
```diff
@@ -963,8 +963,8 @@
   ; Use the already precomputed user->usage map and just aggregate by users to get pool usage.
   (let [pool-usage (reduce (partial merge-with +) (vals user->usage))]
     (->> queue
-         (filter-based-on-pool-quota pool-quota pool-usage)
-         (filter-based-on-user-quota user->quota user->usage))))
+         (filter-based-on-user-quota user->quota user->usage)
+         (filter-based-on-pool-quota pool-quota pool-usage))))
```
Comment (dposada) [inline scheduler/src/cook/tools.clj:967]: Could we add a unit or integration test?

Comment (scrosby) [reply scheduler/src/cook/tools.clj:967]: Done. Currently running through a just-in-case jenkins run.

File: scheduler/test/cook/test/tools.clj [hunk 2]
```diff
@@ -791,6 +791,27 @@
...
 
+(deftest test-filter-pending-jobs-for-quota
+  (let [test-user-1 "john"
+        test-user-2 "bob"
+        user->usage {test-user-1 {:count 1, :cpus 1, :mem 1} test-user-2 {:count 1, :cpus 1, :mem 1}}
+        pool-quota {:count 4, :cpus 100, :mem 100}
+        job-user (atom 0)
+        make-job (fn [user]
+                   (swap! job-user inc)
+                   {:db/id @job-user
```
Comment (dposada) [inline scheduler/test/cook/test/tools.clj:802]: Do we actually need the `:db/id` for this test? Seems like we could get rid of it and the `job-user` `atom`.

File: scheduler/test/cook/test/tools.clj [hunk 3]
```diff
@@ -791,6 +791,27 @@
...
+                    :job/user user
+                    :job/resource [{:resource/type :cpus, :resource/amount 1}
+                                   {:resource/type :mem, :resource/amount 1}]})
+        queue [(make-job test-user-1) (make-job test-user-1) (make-job test-user-1) (make-job test-user-2)]
+        user->quota {test-user-1 {:count 2, :cpus 100, :mem 100} test-user-2 {:count 2, :cpus 100, :mem 100}}
+        [job-1 job-2 job-3 job-4] queue]
+    ; If we filtered pool quota first, only the first two jobs in the
+    ; queue would be seen by the user quota and we'd only launch job-1.
```
Comment (dposada) [inline scheduler/test/cook/test/tools.clj:810]: I don't understand this comment -- each of the 4 jobs in the queue is asking for 1 cpu and 1 mem, and the pool quota is 4 jobs / 100 cpus / 100 mem. So why would the pool quota filter to only the first 2 jobs?

Comment (scrosby) [reply scheduler/test/cook/test/tools.clj:810]: There's two jobs already running (user->usage)

File: scheduler/test/cook/test/tools.clj (hunk 2 above)
Comment (scrosby) [reply scheduler/test/cook/test/tools.clj:802]: Nope. Good find.

File: scheduler/test/cook/test/tools.clj (hunk 3 above)
Comment (dposada) [reply scheduler/test/cook/test/tools.clj:807]: ah, thanks!

**Pull request merged**